src/cls/async-hooks.ts

```diff
@@ -64,51 +98,75 @@ export class AsyncHooksCLS<Context extends {}> implements CLS<Context> {
   }
 
   enable(): void {
-    this.currentContext.value = this.defaultContext;
+    this.contexts = {};
     this.hook.enable();
     this.enabled = true;
   }
 
   disable(): void {
-    this.currentContext.value = this.defaultContext;
+    this.contexts = {};
     this.hook.disable();
     this.enabled = false;
   }
 
   getContext(): Context {
-    return this.currentContext.value;
+    // We don't store this.defaultContext directly in this.contexts.
+    // Getting undefined when looking up this.contexts means that it wasn't
+    // set, so return the default context.
+    const current = this.contexts[this.ah.executionAsyncId()];
+    return current ? current.value : this.defaultContext;
   }
 
   setContext(value: Context): void {
-    this.currentContext.value = value;
+    this.contexts[this.ah.executionAsyncId()].value = value;
   }
 
   runWithNewContext<T>(fn: Func<T>): T {
-    const oldContext = this.currentContext.value;
-    this.currentContext.value = this.defaultContext;
+    // Run fn() so that any AsyncResource objects that are created in
+    // fn will have the context set by this.setContext.
+    const id = this.ah.executionAsyncId();
+    const oldContext = this.contexts[id];
+    // Reset the current context. This prevents this.getContext from returning
+    // a stale value.
+    this.contexts[id] = {value: this.defaultContext};
     try {
       return fn();
     } finally {
-      this.currentContext.value = oldContext;
+      // Revert the current context to what it was before any calls to
+      // this.setContext from within fn.
+      this.contexts[id] = oldContext;
     }
   }
 
   bindWithCurrentContext<T>(fn: Func<T>): Func<T> {
-    if ((fn as ContextWrapped<Func<T>>)[WRAPPED] || !this.currentContext) {
+    const outerId = this.ah.executionAsyncId();
+    // Return if we have already wrapped the function, or there is no current
+    // context to bind.
+    if ((fn as ContextWrapped<Func<T>>)[WRAPPED] || !this.contexts[outerId]) {
       return fn;
     }
-    const current = this.currentContext;
-    const boundContext = this.currentContext.value;
+    const that = this;
+    // Capture the context of the current AsyncResource.
+    const boundContext = this.contexts[outerId];
```

Comment:
Nit. It'd be better to move this line to before line 145, so that `this.contexts[outerId]` can be called only once.